### PR TITLE
Document process of adding an operator

### PR DESCRIPTION
**Pull request by jackm321**
*Description*:
Add a file documenting the basic steps for implementing a new operator.

*Testing*:
Look at markdown [preview](https://github.com/jackm321/glow/blob/new_operator_doc/docs/NewOperators.md)

*Documentation*:
yes

### CODE REVIEW

File: docs/NewOperators.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
+## How to implement a new operator in Glow
+
+> Example PR: https://github.com/pytorch/glow/pull/2334
+
+### Implementation
+* Create a new glow high level IR node in `ClassGen/NodeGen.cpp`.
```
Comment (rdzhabarov) [inline docs/NewOperators.md]: This is not always the case. Sometimes it makes sense to do direct lowering, e.g., FCTransposed.

File: docs/NewOperators.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+* Implement the `verify()` method for the new node in `Graph/Nodes.cpp`.
+* Implement a node creation method in `Graph/Graph.cpp`.
+* Implement logic to load models that contains the operator in `Importer/Caffe2ModelLoader.cpp` or `Importer/ONNXModelLoader.cpp` depending on which type of model the operator comes from. Add the operator to `Importer/CommonOperatorLoader.h` instead if the loading logic can be shared between Caffe2 and ONNX.
+* Create a new glow low level IR instruction in `ClassGen/InstrGen.cpp`
+* Implement the interpreter implementation of the operator in `Interpreter/InterpreterNodes.cpp`
```
Comment (rdzhabarov) [inline docs/NewOperators.md]: I'd also specify that there are several backends which needs to have operator implementation.
Interpreter, CPU, OpenCL. Having CPU implementation is highly encouraged as interpreter is basically a thing to play with but really hard to use for ONNXIFI and other scenarious.

File: docs/NewOperators.md (hunk 1 above)
Comment (jfix71) [reply docs/NewOperators.md]: super nit: Can we always capitalize "Glow" 😄

File: docs/NewOperators.md [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+### Testing
+* Test the operator implementation by writing unit tests in `unittests/OperatorTest.cpp`.
+* Test the operator importing logic by creating a new model in `tests/models/caffe2Models` or `tests/models/onnx2Models` directory then write unit tests in `unittests/Caffe2ImporterTest.cpp` or `tests/unittests/OnnxImporterTest.cpp` using that model.
+
+### PR
+* Create a pull request and prefix the title with "[New operator]".
```
Comment (rdzhabarov) [inline docs/NewOperators.md]: Could you also link the general guidance on how to create Glow PR here?

File: docs/NewOperators.md [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+* Create a new glow high level IR node in `ClassGen/NodeGen.cpp`.
+* Implement the `verify()` method for the new node in `Graph/Nodes.cpp`.
+* Implement a node creation method in `Graph/Graph.cpp`.
+* Implement logic to load models that contains the operator in `Importer/Caffe2ModelLoader.cpp` or `Importer/ONNXModelLoader.cpp` depending on which type of model the operator comes from. Add the operator to `Importer/CommonOperatorLoader.h` instead if the loading logic can be shared between Caffe2 and ONNX.
+* Create a new glow low level IR instruction in `ClassGen/InstrGen.cpp`
+* Implement the interpreter implementation of the operator in `Interpreter/InterpreterNodes.cpp`
+
+### Testing
+* Test the operator implementation by writing unit tests in `unittests/OperatorTest.cpp`.
+* Test the operator importing logic by creating a new model in `tests/models/caffe2Models` or `tests/models/onnx2Models` directory then write unit tests in `unittests/Caffe2ImporterTest.cpp` or `tests/unittests/OnnxImporterTest.cpp` using that model.
```
Comment (rdzhabarov) [inline docs/NewOperators.md]: provide guidance on how to handle scenarios when there is a common implementation.
Like having `tests/models/caffe2Models` is enough for test coverage.

File: docs/NewOperators.md [hunk 5]
```diff
@@ -0,0 +1,18 @@
+## How to implement a new operator in Glow
+
+> Example PR: https://github.com/pytorch/glow/pull/2334
+
+### Implementation
+* Create a new glow high level IR node in `ClassGen/NodeGen.cpp`.
+* Implement the `verify()` method for the new node in `Graph/Nodes.cpp`.
+* Implement a node creation method in `Graph/Graph.cpp`.
+* Implement logic to load models that contains the operator in `Importer/Caffe2ModelLoader.cpp` or `Importer/ONNXModelLoader.cpp` depending on which type of model the operator comes from. Add the operator to `Importer/CommonOperatorLoader.h` instead if the loading logic can be shared between Caffe2 and ONNX.
+* Create a new glow low level IR instruction in `ClassGen/InstrGen.cpp`
```
Comment (jfix71) [inline docs/NewOperators.md]: Many Instructions use `autoIRGen()` in InstrGen, however not all can. So perhaps specify here that if your node cannot use `autoIRGen()` then you will need to add a new case to IRGen to handle it.

File: docs/NewOperators.md [hunk 6]
```diff
@@ -0,0 +1,18 @@
+## How to implement a new operator in Glow
+
+> Example PR: https://github.com/pytorch/glow/pull/2334
+
+### Implementation
+* Create a new glow high level IR node in `ClassGen/NodeGen.cpp`.
+* Implement the `verify()` method for the new node in `Graph/Nodes.cpp`.
+* Implement a node creation method in `Graph/Graph.cpp`.
+* Implement logic to load models that contains the operator in `Importer/Caffe2ModelLoader.cpp` or `Importer/ONNXModelLoader.cpp` depending on which type of model the operator comes from. Add the operator to `Importer/CommonOperatorLoader.h` instead if the loading logic can be shared between Caffe2 and ONNX.
```
Comment (rdzhabarov) [inline docs/NewOperators.md]: I'd mention that we need to try catch as many error cases in the loader as we could (so bunch of error checks here at the model loading time). Especially important for ONNXIFI flow when we reject operator support.

File: docs/NewOperators.md [hunk 7]
```diff
@@ -0,0 +1,24 @@
+## How to implement a new operator in Glow
+
+*Note: It's best to avoid implementing new Glow IR nodes and instructions. If support for a new operator is needed, it's best to first check to see if this operator can be implemented using an existing glow IR node or nodes. For example FCTransposed can be implemented using a regular FC node. If this is the case then the operator can be implemented by instantiating existing glow IR nodes in the loader or by simply creating a new node creation method in `Graph/Graph.cpp` which would create the sequence of operators that implements the new operator. Only in the event that this is not possible should new Glow IR nodes/instructions be created.*
+
+> Example PR: https://github.com/pytorch/glow/pull/2334
+
+### Implementation
+#### High level IR
+* Create a new Glow high level IR node in `ClassGen/NodeGen.cpp`. Run `ninja all` to generate the node.
```
Comment (beicy) [inline docs/NewOperators.md]: My understanding is, we can run "ninja all" for release/debug build, and check the AutoGenNodes.h/cpp file to make sure the node is added?  If so, maybe we can explain more here :)

File: docs/NewOperators.md [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+* Implement the `verify()` method for the new node in `Graph/Nodes.cpp`.
+* Implement a node creation method in `Graph/Graph.cpp`.
+* Implement logic to load model that contains the operator in `Importer/Caffe2ModelLoader.cpp` or `Importer/ONNXModelLoader.cpp` depending on which type of model the operator comes from. Add the operator to `Importer/CommonOperatorLoader.h` instead if the loading logic can be shared between Caffe2 and ONNX. Add as much validation logic as possible here in the loader for the operator because it's crucial to catch errors at this stage. Once the operator is loaded, it is assumed that Glow will be able to successfully run the operator so any issues must be caught here.
+#### Low level IR
+* Create a new Glow low level IR instruction in `ClassGen/InstrGen.cpp`.
+* If a custom translation from high level IR node to low level IR instruction is necessary then implement a new case in `IR/IRGen` otherwise just use `autoIRGen()` in the definition of the new IR instruction in `ClassGen/InstrGen.cpp`.
```
Comment (rdzhabarov) [inline docs/NewOperators.md]: when you specify autoIRGen() how would the mapping happen to the Node generated in the NodeGen. Can you add that?

File: docs/NewOperators.md [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+* If the new instruction can be natively supported on some backends but not all backends then consider implementing a lowering in `Optimizer/Lower.cpp` to lower the high level IR node to other high level IR nodes instead of generating the low level IR instruction for the backends which cannot natively implement the new instruction.
```
Comment (rdzhabarov) [inline docs/NewOperators.md]: New node, just to highlight that lowering happens on a graph level.

**Pull request merged**